rfcs/0159-private-task-logs.md

```diff
@@ -0,0 +1,88 @@
+
+This RFC proposes making this behavior optional, with private logs as an alternative.
+Such logs would require credentials with specific scopes to read.
+
+## Motivation
+
+Even in open-source development, secrecy can be important.
+For example, CI for fixes to security issues can reveal those security issues if logs are public.
+
+Private-log capability is also a common request from outside parties interested in deploying Taskcluster.
+
+# Details
+
+## Current Log Implementation
+
+The current [convention](https://docs.taskcluster.net/docs/manual/design/conventions/task-logs) is to name the main task log `public/logs/live.log`, and that fetching this artifact may -- if the worker supports it -- result in receiving log lines as they are produced.
+
+The Taskcluster UI, and other consumers of task logs such as Treeherder, incorporate this convention and look for logs at that artifact name.
+
+Workers that cannot do "live logging" simply create an S3 artifact with this name.
+Workers that *can* do live logging use the following process.
+At the beginning of the task, they create a Reference artifact named `public/logs/live.log` containing a URL that will connect to the worker (usually via [Websocktunnel](https://github.com/taskcluster/taskcluster/tree/master/tools/websocktunnel#readme)) to get log lines as they are produced.
+When the task is complete, they create a new S3 artifact containing the whole log, named `public/logs/live_backing.log`.
+They then modify the URL in the `public/logs/live.log` artifact (using a [special case in the `createTask` API](https://docs.taskcluster.net/docs/reference/platform/queue/api#createArtifact)) to point to the URL for the `getArtifact` method for the `public/logs/live_backing.log` artifact (`<rootUrl>/api/queue/v1/task/<taskId>/runs/<runId>/artifacts/public/logs/live_backing.log`).
+
+Reference artifacts correspond to HTTP 303 redirects, and HTTP clients do not re-authenticate when following a redirect.
+So the fetch of `live_backing.log` is performed without Taskcluster credentials, even if the initial call to `getArtifact` used credentials.
+If `live_backing.log` is always public, this is not a problem, but if it is private then this may result in additional complexity for clients wishing to access task logs.
+
+## New Log Implementation
+
+The updated convention is, briefly, to make the artifact name containing logs configurable on a per-task basis, so that users may choose a `public/` or non-public name; and to introduce a new artifact type that will avoid the issues around authentication and redirects.
+
+The convention for task-related conventions is to put information in `task.extra`, using sub-objects to group related configuration.
+In the new design, the log artifact name is found at `task.extra.logs.taskLogArtifact`, with a default value of `public/logs/live.log` for compatibility.
```

Comment:
Defaulting to public will make migrating easier but will be a potentially surprising default later in time. I am thinking it would be better to ensure that current tasks we want to be public all update their `task.extra` before we land this change and default to private logs.

The nice thing about this is that the fallout if we make some tasks logs private by accident isn't that bad, we just use our scopes to go get the artifact for the interested party and update their config so that it Just Works in the future.

Reply:
Oh, I see this line is already in discussion above. I'll resolve this and add this as a comment to the other discussion.